### DISCUSSION TITLE
TypeScript: Help with "advanced" usage of helper include functions

### DISCUSSION THREAD

**Discussion by JakubKoralewski** [prisma/prisma#4933] [Q&A] [answered]
I want to write a helper function which depending on the `GetOptional` enum's value will either provide some given `SOMERELATEDFIELDWhereInput` if the `GetOptional.All` and when `GetOptional.JustIds` is set it would include just the `id` field of the given related table. If `GetOptional.Nothing` is set it would return `undefined` meaning the `include` parameter would be undefined, so the field would not be fetched. The result of that function would be provided to the `include` property of a find query.

<details>
<summary>
This is how I attempted to write this:
</summary>

```ts
export enum GetOptional {
	Nothing,
	JustIds,
	All
}
export function ifNotNothingWithOvewrite<
// Ideally the type paramater would not the `Delegate` but the Model itself. 
	Delegate extends Prisma.MeetingDelegate | Prisma.FriendRequestsDelegate
>(
	optional: GetOptional
): <
	T extends Parameters<Delegate['findMany']>[0],
	P extends T['select'] = T['select']
>(ifNotNothing: T, ifJustIdsOverwrite: P) =>
   // what to do here
 // EDIT: I misunderstood a very basic idea that the provided enum is a 
// variable not a compile-time variable so it doesn't make sense to 
// check the output at compile time
	typeof optional extends GetOptional.All ?
		T :
		typeof optional extends GetOptional.JustIds ?
			ReturnType<Delegate['findMany']> :
		undefined
{
	const isNotNothing = optional !== GetOptional.Nothing
	const isJustIds = optional === GetOptional.JustIds
	return (ifNotNothing, ifJustIdsOverwrite) => isNotNothing ?
		(
			isJustIds && ifJustIdsOverwrite ?
				{ ...ifNotNothing, ...ifJustIdsOverwrite} :
				ifNotNothing
		) : undefined
}
```
</details>
But I don't know how to do this because somehow I would like to be able to say that the function returns just the fields provided by the return of the function. How could I do this?

<details>
<summary>
I later would like to use it as follows
</summary>

```ts
	const ifIncludeMeetingsNotNothing = ifNotNothingWithOverwrite(includeMeetings)
	const ifIncludeFriendsNotNothing = ifNotNothingWithOverwrite(includeFriends)
	const friendRequestsSent =
		ifIncludeFriendsNotNothing<FindManyFriendRequestsArgs>(
			{
				where: {
					requesteeId: sessionUserId,
					friendshipType: FriendshipType.ACCEPTED
				}
			},
			{
				requesteeId: true,
				requesterId: true
			}
		)
	const friendRequestsReceived =
		ifIncludeFriendsNotNothing<FindManyFriendRequestsArgs>(
			{
				where: {
					requesteeId: sessionUserId,
					friendshipType: FriendshipType.ACCEPTED
				}
			},
			{
				requesteeId: true,
				requesterId: true
			}
		)
	const meetingsCreated =
		ifIncludeMeetingsNotNothing<FindManyMeetingArgs>(
			{
				where: {
					state: MeetingState.ACCEPTED,
					inviteeId: sessionUserId
				}
			},
			{
				id: true
			}
		)
	const meetingsReceived =
		ifIncludeMeetingsNotNothing<FindManyMeetingArgs>(
			{
				where: {
					state: MeetingState.ACCEPTED,
					inviterId: sessionUserId
				}
			},
			{
				id: true
			}
		)

	const user = await prisma.user.findUnique({
		where: {
			id: userIdInQuestion
		},
		include: {
			friendRequestsSent,
			friendRequestsReceived,
			meetingsCreated,
			meetingsReceived
		}
	})
```

</details>

<details>
<summary>
For now though since I don't know how to do this the return type of my function is `T | undefined`:
</summary>

```ts
export function ifNotNothingWithOverwrite(
	optional: GetOptional
): <T extends FindManyFriendRequestsArgs | FindManyMeetingArgs,
	P extends T['select'] = T['select']>(ifNotNothing: T, ifJustIdsOverwrite: P) => T | undefined {
	const isNotNothing = optional !== GetOptional.Nothing
	const isJustIds = optional === GetOptional.JustIds
	return (ifNotNothing, ifJustIdsOverwrite) => isNotNothing ?
		(
			isJustIds && ifJustIdsOverwrite ?
				{ ...ifNotNothing, ...ifJustIdsOverwrite } :
				ifNotNothing
		) : undefined
}
```

</details>

But this leads to problems because then TypeScript doesn't know what is the return value of the `findUnique` query. I'd appreciate any help.

EDIT: Basically what I'd like to happen is that since the include is possibly undefined:
![obraz](https://user-images.githubusercontent.com/43069023/104104084-37d9c700-52a6-11eb-8aab-0080fef0ee32.png)

the `user .friendRequestsSent` should also be possibly undefined
![obraz](https://user-images.githubusercontent.com/43069023/104104095-42945c00-52a6-11eb-8a7d-7cfa474dee0b.png)

Note:
I know this may seem like reimplementing GraphQL, so if there was a better way to encapsulate this pattern I'd appreciate any tips, as I'm not familiar with GraphQL and decided on a REST-like API and probably don't want to change now.

**Comment by timsuchanek** [comment, marked as answer]
Thanks @JakubKoralewski for asking!
What you are doing here might indeed be easier with GraphQL.
From what I understand, you're passing in a `boolean` into the `include` parameter. And you would like that to end up in `friendRequestsSent`. Right now the `include` API only works with "it's there or it isn't". But we should also make these cases work.
We have this issue as a tracking issue for this: https://github.com/prisma/prisma/issues/5003
Until then, you can by hand add a `| undefined` in there as a hack, which obviously is not nice, but kinda works.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
